pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -1392,6 +1452,157 @@ func (d *pmemCSIDeployment) getNodeRegistrarContainer() corev1.Container {
+		APIGroup: "rbac.authorization.k8s.io",
+		Kind:     "ClusterRole",
+		Name:     d.NodeSetupClusterRoleName(),
+	}
+}
+
+func (d *pmemCSIDeployment) getNodeSetupDaemonSet(ds *appsv1.DaemonSet) {
+	directoryOrCreate := corev1.HostPathDirectoryOrCreate
+
+	if ds.Labels == nil {
+		ds.Labels = map[string]string{}
+	}
+	ds.Labels["app.kubernetes.io/name"] = "pmem-csi-node-setup"
+	ds.Labels["app.kubernetes.io/part-of"] = "pmem-csi"
+	ds.Labels["app.kubernetes.io/component"] = "node-setup"
+	ds.Labels["app.kubernetes.io/instance"] = d.Name
+
+	spec := &ds.Spec
+	spec.Selector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"app.kubernetes.io/name":     "pmem-csi-node-setup",
+			"app.kubernetes.io/instance": d.Name,
+		},
+	}
+	spec.Template.ObjectMeta.Labels = joinMaps(
+		d.Spec.Labels,
+		map[string]string{
+			"app.kubernetes.io/name":      "pmem-csi-node-setup",
+			"app.kubernetes.io/part-of":   "pmem-csi",
+			"app.kubernetes.io/component": "node-setup",
+			"app.kubernetes.io/instance":  d.Name,
+			"pmem-csi.intel.com/webhook":  "ignore",
+		})
+	podSpec := &ds.Spec.Template.Spec
+	podSpec.ServiceAccountName = d.NodeSetupServiceAccountName()
+	podSpec.Tolerations = []corev1.Toleration{
+		{
+			// Allow this pod to run on a master node.
+			Key:    "node-role.kubernetes.io/master",
+			Effect: "NoSchedule",
+		},
+	}
+	podSpec.NodeSelector = map[string]string{
+		d.Name + "/convert-raw-namespaces": "force",
+	}
+	podSpec.Containers = []corev1.Container{
+		d.getNodeSetupContainer(),
+	}
+	podSpec.Volumes = []corev1.Volume{
+		{
+			Name: "dev-dir",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: "/dev",
+					Type: &directoryOrCreate,
+				},
+			},
+		},
+		{
+			Name: "sys-dir",
+			VolumeSource: corev1.VolumeSource{
+				HostPath: &corev1.HostPathVolumeSource{
+					Path: "/sys",
+					Type: &directoryOrCreate,
+				},
+			},
+		},
+	}
+}
+
+func (d *pmemCSIDeployment) getNodeSetupContainer() corev1.Container {
+	true := true
+	root := int64(0)
+	c := corev1.Container{
+		Name:            "pmem-driver",
+		Image:           d.Spec.Image,
+		ImagePullPolicy: d.Spec.PullPolicy,
+		Command:         d.getNodeSetupCommand(),
+		Env: []corev1.EnvVar{
+			{
+				Name: "KUBE_NODE_NAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						APIVersion: "v1",
+						FieldPath:  "spec.nodeName",
+					},
+				},
+			},
+			{
+				Name:  "TERMINATION_LOG_PATH",
+				Value: "/tmp/termination-log",
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      "dev-dir",
+				MountPath: "/dev",
+			},
+			{
+				Name:      "sys-dir",
+				MountPath: "/sys",
+			},
+			{
+				Name:      "sys-dir",
+				MountPath: "/host-sys",
+			},
+		},
+		SecurityContext: &corev1.SecurityContext{
+			Privileged: &true,
+			// Node setup must run as root user
+			RunAsUser: &root,
+		},
+		TerminationMessagePath: "/tmp/termination-log",
```

Comment:
Can you please also set  containers `TerminationMessagePolicy` to the default value of `corev1.TerminationMessageReadFile`. Otherwise, it ends up patching the deamonset on every reconcile.

Reply:
This really is impossible to get right without a test.

Anyway, added.

Reply:
> This really is impossible to get right without a test.

Yes, true.